sql/hash/hash.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2025 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hash
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/cespare/xxhash/v2"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+)
+
+var digestPool = sync.Pool{
+	New: func() any {
+		return xxhash.New()
+	},
+}
+
+// HashOf returns a hash of the given value to be used as key in a cache.
+func HashOf(ctx *sql.Context, sch sql.Schema, row sql.Row) (uint64, error) {
+	hash := digestPool.Get().(*xxhash.Digest)
+	hash.Reset()
+	defer digestPool.Put(hash)
+	for i, v := range row {
+		if i > 0 {
+			// separate each value in the row with a nil byte
+			if _, err := hash.Write([]byte{0}); err != nil {
+				return 0, err
+			}
+		}
+
+		v, err := sql.UnwrapAny(ctx, v)
+		if err != nil {
+			return 0, fmt.Errorf("error unwrapping value: %w", err)
+		}
+
+		// TODO: we may not always have the type information available, so we check schema length.
+		//   Then, defer to original behavior
+		if i >= len(sch) || v == nil {
+			_, err := fmt.Fprintf(hash, "%v", v)
+			if err != nil {
+				return 0, err
+			}
+			continue
+		}
+
+		switch typ := sch[i].Type.(type) {
+		case types.ExtendedType:
+			// TODO: this should use types.ExtendedType.SerializeValue, but there are some doltgres conversion issues
+			//   we need to address. Resort to old behavior for now.
```

Comment:
I'd change this to say something more along the lines of:
```
Doltgres follows Postgres conventions which don't align with the expectations of MySQL, so we're using the old (probably incorrect) behavior for now
```
`SerializeValue` isn't guaranteed to create comparably-unique values, which are different than truly unique values. For example, `1.0` and `1.00` are the same value via comparison, but they may be serialized differently depending on the type and whether the number of significant digits is important. In Postgres, we'd use the `=` operator, but GMS/MySQL doesn't have a concept of bespoke equality functions like that, hence why this doesn't really work.